### PR TITLE
Make std::string conform to CustomDebugStringConvertible

### LINKED ISSUE #62678: [cxx-interop] `std::string` should provide a Swift `debugDescription` property
When debugging Swift code that uses C++ `std::string`, it would be convenient to see a nice and human-readable representation of `std::string` in the debugger.

`std::string` could be conformed to `CustomDebugStringConvertible` protocol by adding an extension in the `std` overlay (see `swift/stdlib/public/Cxx/std/String.swift`).

### PR DESCRIPTION
**Pull request by whiteio**
It'd be useful for `std::string` to conform to `CustomDebugStringConvertible` to see a human readable representation of `std::string` in the debugger.

Resolves #62678

### CODE REVIEW

**Review by zoecarver — APPROVED** [review]
Excellent idea. This LGTM. Let's wait for someone from the standard library to approve it before it lands, though.

**Review by egorzhdan — COMMENTED** [review]
Thanks!
One interesting detail here is that `Swift.String` has to be a valid UTF-8 string, while `std::string` doesn't necessarily have to be. Since `debugDescription` is called by the debugger whenever you hit a breakpoint, it's important that we don't crash while generating a debug description for any kind of strings. Could you please add a test to make sure we don't crash for invalid UTF-8?

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by whiteio** [comment]
> Thanks! One interesting detail here is that `Swift.String` has to be a valid UTF-8 string, while `std::string` doesn't necessarily have to be. Since `debugDescription` is called by the debugger whenever you hit a breakpoint, it's important that we don't crash while generating a debug description for any kind of strings. Could you please add a test to make sure we don't crash for invalid UTF-8?

@egorzhdan Thanks for the feedback. Yes I've had a look at this and I have a question. What would be the right way to initilalize an instance of `std.string` with an invalid UTF-8 string? I was trying to get the `std.string` value to be `"\xE1\xC1\xAC"` but I haven't been able to figure out how to set the value to this. 

I tried using `.push_back` on the `std.string` to add each byte of the invalid UTF-8 value but couldn't get that working as the parameter type for `.push_back` was `Int8` which lead it to overflow. Could you point me in the right direction please?

**Comment by whiteio** [comment]
Thanks for the feedback, I got it working and added in a test for invalid UTF-8.

**Review by egorzhdan — APPROVED** [review]
This looks great, thank you @whiteio!

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
The test failure on macOS doesn't seem related to the patch.

On Linux it looks like the `std::string` initializer is imported differently:
```
/home/build-user/swift/test/Interop/Cxx/stdlib/overlay/std-string-overlay.swift:49:50: error: cannot convert value of type 'Int' to expected argument type 'std.__CxxTemplateInstSaIcE'
  let cxx3: std.string = std.string(cxx3Pointer, 3)
```

**Comment by whiteio** [comment]
@swift-ci Please smoke test Linux platform

**Comment by whiteio** [comment]
@egorzhdan I've fixed the broken test, it's now passing on macOS and Linux. Could you re-run the smoke test please? Sorry about the delay.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@whiteio sorry for the delay, could you please rebase the PR on top of the main branch instead of having a merge commit?

**Comment by whiteio** [comment]
@egorzhdan Yes, no problem. I've rebased it instead of creating a merge commit.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: stdlib/public/Cxx/std/String.swift
Comment (lorentey) [inline stdlib/public/Cxx/std/String.swift:28]: Any reason not to _also_ add a `CustomStringConvertible` conformance?

```swift
extension std.string: CustomStringConvertible {
  public var description: String {
    String(self)
  }
}
```

Unfortunately, once we ship these types, it seems unlikely we'll ever be able to add any new conformances to them to pre-existing protocols, so it might be wise to consider doing this for _all_ newly imported types, as a matter of course.

The same gotcha applies to `Equatable`, `Comparable` and `Hashable` -- if there is a nonzero chance we want to make any these C++ types conform to those later, then we need to add those conformances up front. (I have _immense_ regrets about some basic stdlib types not conforming to these.)

**Review by lorentey — APPROVED** [review]
Approved 100%, with an important worry about the need to add even more explicit conformances up front.

File: stdlib/public/Cxx/std/String.swift
Comment (whiteio) [reply stdlib/public/Cxx/std/String.swift:28]: Good to know, it does seem like I good idea to have 'CustomStringConvertible' conformance too. I'll create another PR to add that conformance in.

**Pull request merged**
